### ISSUE TITLE
TUF-style project metadata validation

### ISSUE THREAD

**Issue by kim** [radicle-dev/radicle-link#13]
This could be a pure state machine for starters.

**Comment by FintanH** [comment]
In the spec it says:

>The co-maintainer's views of the project metadata file are fetched.
[...] 
If no quorum of maintainers can be resolved, the project is invalid.

Is the quorum in regards to the projects agreeing on the initial revision? Because step 3 also talks about checking the signatures for a quorum?

**Comment by kim** [comment]
The description assumes you are connected to some random peer from which you clone a project of (potentially) some other peer. You find the initial project metadata, which is signed by “other peer” - this is the signature you trust (obviously, if you want to be sure, you should have obtained “other peer”s public key via some other channel).

Now, if “other peer” says in their metadata file that there are more maintainers (which could also be other devices of “other peer”), we will need to find the metadata files of those maintainers. If less than a quorum can be found, there’s no way we can reach a quorum of signatures in step 3, so we can stop here.

Does that answer your question?

**Comment by FintanH** [comment]
Ah, I see. So if we say in pseudo-code:

```haskell
let other_peers = fetch(peer.maintainers)
if quorum > len(other_peers) / 2
  then fail QuorumNotReached
  else step3
```

Is that right? Essentially you bail if you can't fetch enough of the maintainers?

**Comment by kim** [comment]
Yeap. Except I think the angle bracket points the wrong way 🤓

**Comment by FintanH** [comment]
Oh sí, I copied the spec but that `>` was the len calculation in that case 😁

**Comment by FintanH** [comment]
The thoughts are really springing at me this weekend, so another question around the same mechanism.

I think there might be an edge case here with fetching a quorum. What happens if you try to fetch the history of each maintainer but less than a quorum still exists, i.e. the co-maintainers have permanently deleted their repository (maybe they don't care about the project anymore)?

I believe that would put the project in an unverifiable state, permanently.

**Comment by kim** [comment]
For an observer (i.e. non-maintainer), the algorithm demands the state to be reset to the latest “good” revision if verification fails. This means that maintainers are free to rewrite history, as long as there is a valid transition from the latest good state (yay for not being a blockchain!).

Bootstrapping is a bit peculiar: if a founder names co-maintainers in the initial revision, those should better be online at the time the project is published, because, since there is no valid state to reset to, nobody else will replicate it. Or else, don’t name co-maintainers initially, but in a subsequent revision.

I’m not sure what you mean by “deleted”: to reach a valid state where A, B and C are maintainers, all their repos are available on the network. Now what happens if A wants to make a change or sign off a new HEAD, but B and C are no longer active? Well yeah, the project is screwed then - consider carefully whom you trust! 

In the case B and C having lost access to their keys, there are ways to perform a key recovery procedure, although there still needs to exist a quorum of trusted parties (cf. https://w3c.github.io/did-core/#key-revocation-and-recovery). I’m not sure yet if we need to make this work within `radicle-link`, or could defer to blockchain-based keys and identity management entirely.

Lastly, the question is also what the impact of a bricked project is. It is technically cheap to declare the source tree a fresh project, but you will need to tell everyone about this, which is less cheap.

**Comment by FintanH** [comment]
> the algorithm demands the state to be reset to the latest “good” revision if verification fails

Can we expand on this a bit more? Does this mean the maintainers can go back and edit the keys to better represent the current state of the project?

> I’m not sure what you mean by “deleted”: to reach a valid state where A, B and C are maintainers, all their repos are available on the network.

What I meant was what if B and C decide not to be maintainers and no longer host the code on their machines? But I guess the sentiment is the same: "Well yeah, the project is screwed then - consider carefully whom you trust!"

This is good food for thought, I'll start thinking about the pure state machine some more and see if any other questions fall out :thinking:

**Comment by FintanH** [comment]
> the revision of N-1 is less than the revision of N

Should this be a quorum as well? My thinking is that we have metadata files from `A`, `B`, `C`. `A` and `B` both have revision `2` in the revision that is after the initial commit, whereas `C` says the revision is `3`. We have a quorum from `A` and `B`.

**Comment by kim** [comment]
For posterity: the revision part is a bit iffy still - TUF says:

> The version number of the trusted root metadata file (version N) must be less than or equal to the version number of the new root metadata file (version N+1)

Where I don't quite understand why it can be equal. Technically, we could do without the numerical revision in git, if we'd establish logical ordering by some other means (e.g. additional parents). That may not generalise well, however.

**Comment by kim** [comment]
As @massimiliano-mantione rightfully points out, an update to the metadata would also need to reference the "point in time" after which this update should apply. The initial commit already has the head of the branch you initialised the project from as a parent (this is debatable, we could also enforce that this branch is called "master", or rather, that it is equal to the `default_branch` value). An update would accordingly have two parents: the initial metadata commit, and the head of `default_branch`.

**Comment by kim** [comment]
Closed via #39
